shared/storage.go

```diff
@@ -284,3 +286,125 @@ func (cs byteCachedStore) Get(cacheID, storeID, iValue interface{}) error {
 func NewByteCachedStore(ctx context.Context, cache ReadWritable, store Readable) CachedStore {
 	return byteCachedStore{ctx, cache, store}
 }
+
+// ObjectStore is a store that populates an arbitrary output object on Get().
+type ObjectStore interface {
+	Get(id, value interface{}) error
+}
+
+// ObjectCache is an ObjectStore that also supports Put() for arbitrary id/value
+// pairs.
+type ObjectCache interface {
+	ObjectStore
+	Put(id, value interface{}) error
+}
+
+type jsonObjectCache struct {
+	ctx      context.Context
+	delegate ReadWritable
+}
+
+func (oc jsonObjectCache) Get(id, value interface{}) error {
+	r, err := oc.delegate.NewReadCloser(id)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err := r.Close()
+		if err != nil {
+			logger := oc.ctx.Value(DefaultLoggerCtxKey()).(Logger)
```

Comment:
This is another instance where we should use a fallback/helper (when getting the logger).

Reply:
Will do once #556 lands.